### PR TITLE
NIFI-10676 USE_SPECIFIED_OR_COMPATIBLE_OR_GHOST on flow load from bytes

### PR DESCRIPTION
**Pull request by genehynson**
# Summary

[NIFI-10676](https://issues.apache.org/jira/browse/NIFI-10676) allows NiFi nodes to use compatible flow resources when the specified resource version is not available while the flow definition is being loaded from the cluster connection response.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 8
  - [ ] JDK 11
  - [ ] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

**Comment by michael81877** [comment]
I tested this code change with a modified version of 1.18.0 and 1.19.0-SNAPSHOT with a basic 3-processor flow. 

1. I deployed a 3-node modified 1.18.0 cluster and let it come up gracefully

1. I updated the nifi image version in my deployment to the modified 1.19.0-SNAPSHOT and upgraded each node 1 by 1. Each upgraded node joined the existing cluster gracefully, which definitely wasn't the case prior to this code change.

1. I changed the nifi image version in my deployment back to 1.18.0 and performed another rolling upgrade. Each node again rejoined the cluster gracefully.

This is a huge feature that enables zero downtime rolling upgrades. There may be edge cases, for example with ghost processors, but the flow stabilizes when the upgrade is complete.

There may be an argument to make this configurable via `nifi.properties`, but this LGTM overall +1

**Comment by markap14** [comment]
-1 I don't think this is a change that we want to make at this time. The existing value of USE_SPECIFIED_OR_FAIL is correct here. When we startup in standalone, not part of a cluster, we want to go ahead and startup, even if the instance is missing one of the extensions referenced in the dataflow. When a user looks at the UI, this is made clear by the fact that the processor is "Ghosted" and tooltips, etc. explain exactly what is wrong, and it is easy to address.

However, when a node is part of a cluster, things are a bit more complex. Every node in the cluster is expected to be doing the same thing. This change would result in some nodes in the cluster doing one thing, while other nodes are potentially doing something completely different. For example, if one of 5 nodes is missing an extension, or has an outdated version of the extension, the UI may show that version 1.19 is running. Meanwhile, one node is running version 1.15 of that extension, which behaves differently. This can lead to a great deal of confusion.

It is also worth noting that making this change would not really enable zero downtime rolling upgrades. It would remove one of the roadblocks (with significant downside/tradeoffs). However, there are many other things that would have to be considered. Things like cluster protocols between nodes, the load-balancing protocol, site-to-site protocol, etc.

In short, there are many parts of the codebase that operate on the assumption that nodes are homogenous (with respect to the nifi version). This change would allow that assertion to no longer hold, which can cause a lot of problems, many of which may be very difficult to understand and troubleshoot.

**Comment by michael81877** [comment]
I could be off base here, so please correct me if i'm wrong. I'm just trying to understand the concerns completely.

If i'm not mistaken, this change doesn't affect the startup of a standalone node. That's [handled separately](https://github.com/apache/nifi/blob/a5d93d96a30b3af73146e4b3c2a36a86332bb774/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/StandardFlowService.java#L554-L564) using the same `BundleUpdateStrategy` strategy proposed in this PR. This PR only affects the strategy used when a node joins a cluster.

Additionally, it seems like protocol versions are negotiated for things like [load balancing](https://github.com/apache/nifi/blob/main/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/clustered/server/StandardLoadBalanceProtocol.java#L125) and site-to-site [1, 2]. Wouldn't that prevent any unexpected issues with communications?

In the scenario where one node could be missing or has an outdated version of an extension that other nodes in the cluster have, i believe that would be considered an error by the operator, right? In cases where deployment environments are entirely version controlled, these sorts of issues would seem unlikely to present themselves. Perhaps a flow requires a version of an extension that a new NiFi version doesn't contain, but i would also consider this an error by the operator.

Since we know NiFi is capable of it rolling upgrades, is there an alternative recommended approach?

[1] [socket site-to-site version negotiation](https://github.com/apache/nifi/blob/main/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-site-to-site/src/main/java/org/apache/nifi/remote/protocol/socket/SocketFlowFileServerProtocol.java#L52)
[2] [http site-to-site version negotiation](https://github.com/apache/nifi/blob/main/nifi-commons/nifi-site-to-site-client/src/main/java/org/apache/nifi/remote/client/http/TransportProtocolVersionNegotiator.java)

**Comment by markap14** [comment]
@michael81877  nope you're correct, that this change does not affect the startup for a standalone node. I was merely pointing out that having different logic for standalone and for clustered nodes is very much intentional, not an oversight. The `USE_SPECIFIED_OR_COMPATIBLE_OR_GHOST` strategy is used on startup for standalone nodes. This allows us to automatically upgrade processors from 1.18 to 1.19 when you upgrade, or to startup with a ghosted component if it's not available. But we don't want that behavior when connecting to a cluster. We don't want some nodes ghosting a component while others do not.

You do have a good point on the protocol version negotiation for site-to-site and load balancing. But we don't have the same for the cluster protocol, which would be important.

The scenario where one node could be missing or has an outdated version of an extension is not necessarily an operator error. This will always be the case when doing a rolling upgrade. Some nodes will be on one version of an extension (say 1.19) while others will be on another version of the extension (say 1.20). The new version may have additional properties, or may be missing certain properties. Or Relationships. But even after the upgrade, there could be a mismatch of versions. And yes, that would likely imply operator error. But just because something is operator error doesn't mean that we should simply let it happen and blame the operator. To the extent that's reasonable, we should prevent operators from doing things that we know are wrong.

To be clear here, I'm not saying "No way, let's not allow for rolling upgrades!"
What I am saying is: "The current value is there intentionally, it's not a bug. We don't want to just switch it and assume that no problems will arise. If we want to go down the path of allowing rolling upgrades, we need to do so carefully and think through all of the consequences and make sure that the changes that we're making make sense."

**Comment by genehynson** [comment]
Thanks for providing this insight @markap14 . Originally I thought this was a bug because of the inconsistency of loading the flow definition from a file vs from the connection response - but I understand now that its intentional. 

I've been using this change on a fork of NiFi for 6 months or so now and haven't run into any of the edge cases you've identified but I understand that they can exist. And I certainly don't want to introduce a change that adds unwanted risk to the normal NiFi upgrade process. 

Would you be open to the idea of introducing this change as an opt-in feature that is enabled with a flag in `nifi.properties`? That would allow more folks to test this capability in a "use at your own risk" fashion so we could better understand the impact and what additional changes would be necessary to fully support rolling upgrades in the future.

**Comment by markap14** [comment]
Sorry for the delay responding. I suppose it's reasonable to add it as an option in nifi.properties, provided that we explain that it should be considered a "beta" feature, explain exactly what it does and name it appropriately (i.e., we shouldn't name it something to the effect of 'nifi.allow.rolling.upgrades' because that's not really what it's doing).

**Comment by michael81877** [comment]
Let us know if a property named `nifi.cluster.flow.bundle.update.strategy` is appropriate. I tried to indicate that it is clustering related and specifically about the flow `BundleUpdateStrategy`. The default behavior is the existing `USE_SPECIFIED_OR_FAIL` behavior and it will only change if the property is explicitly set to a different `BundleUpdateStrategy`.

In other words, nothing will change unless you set this new property to a valid `BundleUpdateStrategy`.

**Comment by genehynson** [comment]
@michael81877 can you kick off the CI again?

**Comment by genehynson** [comment]
@markap14 can you kick off the CI again?

**Comment by michael81877** [comment]
@markap14 : Do you see any other changes needed for this PR?

**Comment by michael81877** [comment]
Are there any other changes needed for this PR? Is it possible to get it merged?

**Comment by mh013370** [comment]
I would still like to see this added, but i don't think i can push to the source branch. If @genehynson isn't in a position to make these changes then i'll raise a new PR with the requested changes in 2 days.

**Comment by genehynson** [comment]
I'm not currently available to update this PR! So sorry. @mh013370 feel free to cherry pick and open a new one.

**Comment by exceptionfactory** [comment]
Thanks for the work and discussion on this PR @genehynson and @mh013370. Based on feedback, it sounds like this change could go forward with sufficient updates to the Admin Guide describing the property, the expected behavior, and the experimental nature of the alternative feel value. I'm closing this for now, but feel free to pick this up and reference it when documentation changes are prepared for review.
